torch/csrc/api/src/nn/modules/transformer.cpp

```diff
@@ -223,5 +224,66 @@ Tensor TransformerDecoderLayerImpl::activation(const Tensor& input){
   }
 }
 
+
+// ========================TransformerEncoderImpl=========================
+TransformerEncoderImpl::TransformerEncoderImpl(
+  TransformerEncoderOptions options_) : options(std::move(options_)) {
+  reset();
+}
+
+void TransformerEncoderImpl::reset() {
+  layers = this->register_module("layers", ModuleList());
+  for (int64_t i = 0; i < options.num_layers(); ++i) {
+    layers->push_back(options.encoder_layer()->clone());
+  }
+
+  if (!options.norm().is_empty()) {
+    norm = options.norm().clone();
+    this->register_module("norm", norm.ptr());
+  }
+}
+
+void TransformerEncoderImpl::reset_parameters() {
+  TORCH_CHECK(
+    layers->size() == options.num_layers(),
+    "TransformerEncoder should have", options.num_layers(), " encoder layers, but got ", layers->size());
+
+  size_t num_layers = layers->size();
+  for (size_t i = 0; i < num_layers; ++i) {
+    layers->at<TransformerEncoderLayerImpl>(i).reset_parameters();
+  }
+  // a. No way to know whether module in AnyModule has api to reset_parameters, so replace instead
+  // b. Allow user to add/delete normalization module when reset parameters
```

Comment:
When does `reset_parameters()` get called? is it a user API or something that is used in the framework?

Reply:
As I mentioned before, this is a convenient function to reset parameters within the module. It is not necessary to call it. And I believe, it won't be called in most of the case.

Reply:
Hmm, so the thing I am confused about here is:
- there's no reset_parameters() or _reset_parameters on TransformerEncoderLayer in Python. Some of our Python nn.Modules do have reset_parameters() that re-initializes parameters.

And one potential concern is, let's say you have an optimizer that takes in a list of parameters (disclaimer: I have no clue how optimizers in C++ work). Is the following possible?
- I create a TransformerEncoderLayer in C++
- I add pass the parameters of a TransformerEncoderLayer to an optimizer (in C++)
- then, I call reset_parameters() on TransformerEncoderLayer
- The optimizer holds onto references of the original parameters (and not the new ones!)

Reply:
This is not possible since the C++ module underlying is a shared pointer, and optimizer doesn't clone the module but just shallow copy the module, so this case should be ok.

But the reset_parameters() it self is tricky, I admit that, even in python impl, reset_parameters() has different usage in different modules and not always there and not always get called. There is a PR that trying to standarize it by making reset_parameter() virtual and moved that to base class, but this change doesn't make sense since half of the module doesn't have it and virtual will make extra cost. So currently, reset_parameters() is just an optional functions.